Fetching indexed field in child document panics

The following setup panics.  It does not panic when `reward_id` is removed from the query.  It does not panic when the index is removed from `ActionRequirement`.

Schema:
```
type ActionRequirement @index(unique: true, fields: ["action_id", "reward_id"], name: "actionId_rewardId")  {
  count: Int
  isPreReq: Boolean # @default(false)
  action: Action @relation(name: "actionActionRequirements")
  reward: Reward @relation(name: "rewardActionRequirements")
  createdAt: DateTime # @default(now())
  oldId: String
}

type Reward {
  name: String @index(unique: true) # @unique
  description: String
  value: Int
  limit: Int
  tiers: [RewardTier] @relation(name: "rewardRewardTiers")
  priority: Int
  enabled: Boolean # @default(true)
  actionRequirements: [ActionRequirement] @relation(name: "rewardActionRequirements")
  playerRewards: [PlayerReward] @relation(name: "rewardPlayerRewards")
  rewardGroup: [String] # [ENUM] RewardGroup
  createdAt: DateTime # @default(now())
  oldId: String
}
```
Query:
```
query {
  Reward(filter: {name: {_eq: "FOLLOW_TWITTER"}}) {
    _deleted
    _docID
    id: _docID
    createdAt
    description
    enabled
    limit
    name
    priority
    rewardGroup
    value
    oldId
    actionRequirements {
      id: _docID
      _docID
      action_id
      count
      createdAt
      isPreReq
      oldId
      reward_id
    }
  }}
```
Panic (was reported via http client, so original panic is lost):
> panic: Unclosed iterator at time of Txn.Discard.
 github.com/sourcenetwork/badger/v4.(Txn).Discard
  github.com/sourcenetwork/badger/v4@v4.2.1-0.20231113215945-a63444ca5276/txn.go:524
    github.com/sourcenetwork/badger/v4.(Txn).Commit
      github.com/sourcenetwork/badger/v4@v4.2.1-0.20231113215945-a63444ca5276/txn.go:665
    github.com/sourcenetwork/defradb/datastore/badger/v4.(txn).commit
      github.com/sourcenetwork/defradb/datastore/badger/v4/datastore.go:765
    github.com/sourcenetwork/defradb/datastore/badger/v4.(txn).Commit
      github.com/sourcenetwork/defradb/datastore/badger/v4/datastore.go:761
    github.com/sourcenetwork/defradb/datastore.(txn).Commit
      github.com/sourcenetwork/defradb/datastore/txn.go:106
    github.com/sourcenetwork/defradb/internal/db.(db).ExecRequest
      github.com/sourcenetwork/defradb/internal/db/store.go:38
    github.com/sourcenetwork/defradb/http.(*storeHandler).ExecRequest
      github.com/sourcenetwork/defradb/http/handler_store.go:302
    net/http.HandlerFunc.ServeHTTP
      net/http/server.go:2171

This panic happens in the CI as well:
```
=== FAIL: tests/integration/index TestQueryWithIndexOnManyToOne_MultipleViaOneToMany (0.19s)
Aug 20 01:42:49.902 INF tests.integration  database=badger-in-memory client=go mutationType=collection-save databaseDir="" badgerEncryption=true skipNetworkTests=false changeDetector.Enabled=false changeDetector.SetupOnly=false changeDetector.SourceBranch=develop changeDetector.TargetBranch="" changeDetector.Repository=https://github.com/sourcenetwork/defradb.git
Aug 20 01:42:50.060 INF db Closing DefraDB process...
Aug 20 01:42:50.088 INF db Successfully closed running process
panic: Unclosed iterator at time of Txn.Discard.
	panic: Unclosed iterator at time of Txn.Discard. [recovered]
	panic: Unclosed iterator at time of Txn.Discard.

goroutine 730931 [running]:
testing.tRunner.func1.2({0x6560f60, 0x87e2ff0})
	/opt/hostedtoolcache/go/1.21.3/x64/src/testing/testing.go:1545 +0x366
testing.tRunner.func1()
	/opt/hostedtoolcache/go/1.21.3/x64/src/testing/testing.go:1548 +0x630
panic({0x6560f60?, 0x87e2ff0?})
	/opt/hostedtoolcache/go/1.21.3/x64/src/runtime/panic.go:920 +0x270
github.com/sourcenetwork/badger/v4.(*Txn).Discard(0xc00191ed00)
	/home/runner/go/pkg/mod/github.com/sourcenetwork/badger/v4@v4.2.1-0.20231113215945-a63444ca5276/txn.go:524 +0x1fe
github.com/sourcenetwork/defradb/datastore/badger/v4.(*txn).discard(...)
	/home/runner/work/defradb/defradb/datastore/badger/v4/datastore.go:780
github.com/sourcenetwork/defradb/datastore/badger/v4.(*txn).Discard(0xc0040c3b00, {0x4ed6d5?, 0xc003b13e78?})
	/home/runner/work/defradb/defradb/datastore/badger/v4/datastore.go:776 +0x1c5
github.com/sourcenetwork/defradb/datastore.(*txn).Discard(0xc002798780, {0x8834840, 0xc00210b2f0})
	/home/runner/work/defradb/defradb/datastore/txn.go:125 +0xb5
panic({0x6560f60?, 0x87e2ff0?})
	/opt/hostedtoolcache/go/1.21.3/x64/src/runtime/panic.go:920 +0x270
github.com/sourcenetwork/badger/v4.(*Txn).Discard(0xc00191ed00)
	/home/runner/go/pkg/mod/github.com/sourcenetwork/badger/v4@v4.2.1-0.20231113215945-a63444ca5276/txn.go:524 +0x1fe
github.com/sourcenetwork/badger/v4.(*Txn).Commit(0xc00191ed00)
	/home/runner/go/pkg/mod/github.com/sourcenetwork/badger/v4@v4.2.1-0.20231113215945-a63444ca5276/txn.go:665 +0x17f
github.com/sourcenetwork/defradb/datastore/badger/v4.(*txn).commit(0xc0040c3b00)
	/home/runner/work/defradb/defradb/datastore/badger/v4/datastore.go:765 +0x98
github.com/sourcenetwork/defradb/datastore/badger/v4.(*txn).Commit(0xc0040c3b00, {0x0?, 0x0?})
	/home/runner/work/defradb/defradb/datastore/badger/v4/datastore.go:761 +0x198
github.com/sourcenetwork/defradb/datastore.(*txn).Commit(0xc002798780, {0x8834840, 0xc00210b2f0})
	/home/runner/work/defradb/defradb/datastore/txn.go:106 +0xa9
github.com/sourcenetwork/defradb/internal/db.(*db).ExecRequest(0x0?, {0x8834840, 0xc00210b140}, {0x6fb80d0, 0x5f})
	/home/runner/work/defradb/defradb/internal/db/store.go:38 +0x36f
github.com/sourcenetwork/defradb/tests/integration.executeRequest(0xc00164d8c0, {{0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {0x6fb80d0, 0x5f}, 0xc002763d40, ...})
	/home/runner/work/defradb/defradb/tests/integration/utils.go:1708 +0x372
github.com/sourcenetwork/defradb/tests/integration.performAction(0xc00164d8c0, 0x0?, {0x6b8f380?, 0xc0013a1110?})
	/home/runner/work/defradb/defradb/tests/integration/utils.go:336 +0x1347
github.com/sourcenetwork/defradb/tests/integration.executeTestCase({0x88347d0?, 0xc597fc0}, {0x888a338?, 0xc00d451380?}, {0xc002763dd0, 0x3, 0x3}, {{0x0, 0x0}, {0xc002b3af50, ...}, ...}, ...)
	/home/runner/work/defradb/defradb/tests/integration/utils.go:226 +0x1276
github.com/sourcenetwork/defradb/tests/integration.ExecuteTestCase({0x888a338?, 0xc00d451380}, {{0x0, 0x0}, {0xc002b3af50, 0x5, 0x5}, {0x0, {0x0, 0x0, ...}}, ...})
	/home/runner/work/defradb/defradb/tests/integration/utils.go:180 +0xade
github.com/sourcenetwork/defradb/tests/integration/index.TestQueryWithIndexOnManyToOne_MultipleViaOneToMany(0x0?)
	/home/runner/work/defradb/defradb/tests/integration/index/query_with_relation_filter_test.go:1018 +0xb88
testing.tRunner(0xc00d451380, 0x72573e0)
	/opt/hostedtoolcache/go/1.21.3/x64/src/testing/testing.go:1595 +0x239
created by testing.(*T).Run in goroutine 1
	/opt/hostedtoolcache/go/1.21.3/x64/src/testing/testing.go:1648 +0x82b
```

--- a/internal/db/fetcher/indexer_iterators.go
+++ b/internal/db/fetcher/indexer_iterators.go
@@ -81,6 +81,11 @@ var _ indexIterator = (*indexPrefixIterator)(nil)
 func (iter *indexPrefixIterator) Init(ctx context.Context, store datastore.DSReaderWriter) error {
 	iter.ctx = ctx
 	iter.store = store
+	if iter.resultIter != nil {
+		if err := iter.resultIter.Close(); err != nil {
+			return err
+		}
+	}
 	iter.resultIter = nil
 	return nil
 }

--- a/internal/db/fetcher/indexer_iterators.go
+++ b/internal/db/fetcher/indexer_iterators.go
@@ -64,13 +64,41 @@ type indexIterResult struct {
 	value    []byte
 }
 
-type queryResultIterator struct {
-	resultIter    query.Results
+// indexPrefixIterator is an iterator over index keys with a specific prefix.
+type indexPrefixIterator struct {
 	indexDesc     client.IndexDescription
 	indexedFields []client.FieldDefinition
+	indexKey      core.IndexDataStoreKey
+	matchers      []valueMatcher
+	execInfo      *ExecInfo
+	resultIter    query.Results
+	ctx           context.Context
+	store         datastore.DSReaderWriter
 }
 
-func (iter *queryResultIterator) Next() (indexIterResult, error) {
+var _ indexIterator = (*indexPrefixIterator)(nil)
+
+func (iter *indexPrefixIterator) Init(ctx context.Context, store datastore.DSReaderWriter) error {
+	iter.ctx = ctx
+	iter.store = store
+	iter.resultIter = nil
+	return nil
+}
+
+func (iter *indexPrefixIterator) checkResultIterator() error {
+	if iter.resultIter == nil {
+		resultIter, err := iter.store.Query(iter.ctx, query.Query{
+			Prefix: iter.indexKey.ToString(),
+		})
+		if err != nil {
+			return err
+		}
+		iter.resultIter = resultIter
+	}
+	return nil
+}
+
+func (iter *indexPrefixIterator) nextResult() (indexIterResult, error) {
 	res, hasVal := iter.resultIter.NextSync()
 	if res.Error != nil {
 		return indexIterResult{}, res.Error
@@ -86,44 +114,29 @@ func (iter *queryResultIterator) Next() (indexIterResult, error) {
 	return indexIterResult{key: key, value: res.Value, foundKey: true}, nil
 }
 
-func (iter *queryResultIterator) Close() error {
-	return iter.resultIter.Close()
-}
-
-type eqPrefixIndexIterator struct {
-	queryResultIterator
-	indexKey core.IndexDataStoreKey
-	execInfo *ExecInfo
-	matchers []valueMatcher
-}
-
-func (iter *eqPrefixIndexIterator) Init(ctx context.Context, store datastore.DSReaderWriter) error {
-	resultIter, err := store.Query(ctx, query.Query{
-		Prefix: iter.indexKey.ToString(),
-	})
-	if err != nil {
-		return err
+func (iter *indexPrefixIterator) Next() (indexIterResult, error) {
+	if err := iter.checkResultIterator(); err != nil {
+		return indexIterResult{}, err
 	}
-	iter.resultIter = resultIter
-	return nil
-}
 
-func (iter *eqPrefixIndexIterator) Next() (indexIterResult, error) {
 	for {
-		res, err := iter.queryResultIterator.Next()
+		res, err := iter.nextResult()
 		if err != nil || !res.foundKey {
 			return res, err
 		}
 		iter.execInfo.IndexesFetched++
-		doesMatch, err := executeValueMatchers(iter.matchers, res.key.Fields)
+		didMatch, err := executeValueMatchers(iter.matchers, res.key.Fields)
 		if err != nil {
 			return indexIterResult{}, err
 		}
-		if !doesMatch {
-			continue
+		if didMatch {
+			return res, nil
 		}
-		return res, err
 	}
+}
+
+func (iter *indexPrefixIterator) Close() error {
+	return iter.resultIter.Close()
 }
 
 type eqSingleIndexIterator struct {
@@ -182,7 +195,7 @@ func (iter *inIndexIterator) nextIterator() (bool, error) {
 	}
 
 	switch fieldIter := iter.indexIterator.(type) {
-	case *eqPrefixIndexIterator:
+	case *indexPrefixIterator:
 		fieldIter.indexKey.Fields[0].Value = iter.inValues[iter.nextValIndex]
 	case *eqSingleIndexIterator:
 		fieldIter.indexKey.Fields[0].Value = iter.inValues[iter.nextValIndex]
@@ -236,41 +249,6 @@ func executeValueMatchers(matchers []valueMatcher, fields []core.IndexedField) (
 		}
 	}
 	return true, nil
-}
-
-type scanningIndexIterator struct {
-	queryResultIterator
-	indexKey core.IndexDataStoreKey
-	matchers []valueMatcher
-	execInfo *ExecInfo
-}
-
-func (iter *scanningIndexIterator) Init(ctx context.Context, store datastore.DSReaderWriter) error {
-	resultIter, err := store.Query(ctx, query.Query{
-		Prefix: iter.indexKey.ToString(),
-	})
-	if err != nil {
-		return err
-	}
-	iter.resultIter = resultIter
-
-	return nil
-}
-
-func (iter *scanningIndexIterator) Next() (indexIterResult, error) {
-	for {
-		res, err := iter.queryResultIterator.Next()
-		if err != nil || !res.foundKey {
-			return indexIterResult{}, err
-		}
-		iter.execInfo.IndexesFetched++
-
-		didMatch, err := executeValueMatchers(iter.matchers, res.key.Fields)
-
-		if didMatch {
-			return res, err
-		}
-	}
 }
 
 // checks if the value satisfies the condition
@@ -468,7 +446,7 @@ func (m *anyMatcher) Match(client.NormalValue) (bool, error) { return true, nil 
 func (f *IndexFetcher) newPrefixIndexIterator(
 	fieldConditions []fieldFilterCond,
 	matchers []valueMatcher,
-) (*eqPrefixIndexIterator, error) {
+) (*indexPrefixIterator, error) {
 	keyFieldValues := make([]client.NormalValue, 0, len(fieldConditions))
 	for i := range fieldConditions {
 		if fieldConditions[i].op != opEq {
@@ -488,16 +466,21 @@ func (f *IndexFetcher) newPrefixIndexIterator(
 
 	key := f.newIndexDataStoreKeyWithValues(keyFieldValues)
 
-	return &eqPrefixIndexIterator{
-		queryResultIterator: f.newQueryResultIterator(),
-		indexKey:            key,
-		execInfo:            &f.execInfo,
-		matchers:            matchers,
-	}, nil
+	return f.newQueryResultIterator(key, matchers, &f.execInfo), nil
 }
 
-func (f *IndexFetcher) newQueryResultIterator() queryResultIterator {
-	return queryResultIterator{indexDesc: f.indexDesc, indexedFields: f.indexedFields}
+func (f *IndexFetcher) newQueryResultIterator(
+	indexKey core.IndexDataStoreKey,
+	matchers []valueMatcher,
+	execInfo *ExecInfo,
+) *indexPrefixIterator {
+	return &indexPrefixIterator{
+		indexDesc:     f.indexDesc,
+		indexedFields: f.indexedFields,
+		indexKey:      indexKey,
+		matchers:      matchers,
+		execInfo:      execInfo,
+	}
 }
 
 // newInIndexIterator creates a new inIndexIterator for fetching indexed data.
@@ -537,12 +520,7 @@ func (f *IndexFetcher) newInIndexIterator(
 		indexKey := f.newIndexDataStoreKey()
 		indexKey.Fields = []core.IndexedField{{Descending: f.indexDesc.Fields[0].Descending}}
 
-		iter = &eqPrefixIndexIterator{
-			queryResultIterator: f.newQueryResultIterator(),
-			indexKey:            indexKey,
-			execInfo:            &f.execInfo,
-			matchers:            matchers,
-		}
+		iter = f.newQueryResultIterator(indexKey, matchers, &f.execInfo)
 	}
 	return &inIndexIterator{
 		indexIterator: iter,
@@ -600,12 +578,7 @@ func (f *IndexFetcher) createIndexIterator() (indexIterator, error) {
 	case opIn:
 		return f.newInIndexIterator(fieldConditions, matchers)
 	case opGt, opGe, opLt, opLe, opNe, opNin, opLike, opNlike, opILike, opNILike:
-		return &scanningIndexIterator{
-			queryResultIterator: f.newQueryResultIterator(),
-			indexKey:            f.newIndexDataStoreKey(),
-			matchers:            matchers,
-			execInfo:            &f.execInfo,
-		}, nil
+		return f.newQueryResultIterator(f.newIndexDataStoreKey(), matchers, &f.execInfo), nil
 	}
 
 	return nil, NewErrInvalidFilterOperator(fieldConditions[0].op)

--- a/internal/planner/multi.go
+++ b/internal/planner/multi.go
@@ -1,4 +1,4 @@
-// Copyright 2022 Democratized Data Foundation
+// Copyright 2024 Democratized Data Foundation
 //
 // Use of this software is governed by the Business Source License
 // included in the file licenses/BSL.txt.
@@ -214,22 +214,22 @@ func (s *selectNode) addSubPlan(fieldIndex int, newPlan planNode) error {
 		if err := s.planner.walkAndReplacePlan(s.source, origScan, multiscan); err != nil {
 			return err
 		}
-		// create multinode
-		multinode := &parallelNode{
+		// create parallelNode
+		parallelNode := &parallelNode{
 			p:         s.planner,
 			multiscan: multiscan,
 			docMapper: docMapper{s.source.DocumentMap()},
 		}
-		multinode.addChild(-1, s.source)
+		parallelNode.addChild(-1, s.source)
 		multiscan.addReader()
 		// replace our new node internal scanNode with our new multiscanner
 		if err := s.planner.walkAndReplacePlan(newPlan, origScan, multiscan); err != nil {
 			return err
 		}
 		// add our newly updated plan to the multinode
-		multinode.addChild(fieldIndex, newPlan)
+		parallelNode.addChild(fieldIndex, newPlan)
 		multiscan.addReader()
-		s.source = multinode
+		s.source = parallelNode
 
 	// we already have an existing parallelNode as our source
 	case *parallelNode:

--- a/internal/planner/scan.go
+++ b/internal/planner/scan.go
@@ -1,4 +1,4 @@
-// Copyright 2022 Democratized Data Foundation
+// Copyright 2024 Democratized Data Foundation
 //
 // Use of this software is governed by the Business Source License
 // included in the file licenses/BSL.txt.
@@ -325,35 +325,53 @@ func (p *Planner) Scan(
 type multiScanNode struct {
 	scanNode   *scanNode
 	numReaders int
-	numCalls   int
+	nextCount  int
+	initCount  int
+	startCount int
+	closeCount int
 
-	lastBool bool
-	lastErr  error
+	nextResult bool
+	nextErr    error
+	initErr    error
+	startErr   error
+	closeErr   error
 }
 
 func (n *multiScanNode) Init() error {
-	return n.scanNode.Init()
+	n.countAndCall(&n.initCount, func() {
+		n.initErr = n.scanNode.Init()
+	})
+	return n.initErr
 }
 
 func (n *multiScanNode) Start() error {
-	return n.scanNode.Start()
+	n.countAndCall(&n.startCount, func() {
+		n.startErr = n.scanNode.Start()
+	})
+	return n.startErr
+}
+
+func (n *multiScanNode) countAndCall(count *int, f func()) {
+	if *count == 0 {
+		f()
+	}
+	*count++
+
+	// if the number of calls equals the numbers of readers
+	// reset the counter, so our next call actually executes the function
+	if *count == n.numReaders {
+		*count = 0
+	}
 }
 
 // Next only calls Next() on the underlying
 // scanNode every numReaders.
 func (n *multiScanNode) Next() (bool, error) {
-	if n.numCalls == 0 {
-		n.lastBool, n.lastErr = n.scanNode.Next()
-	}
-	n.numCalls++
+	n.countAndCall(&n.nextCount, func() {
+		n.nextResult, n.nextErr = n.scanNode.Next()
+	})
 
-	// if the number of calls equals the numbers of readers
-	// reset the counter, so our next call actually executes the Next()
-	if n.numCalls == n.numReaders {
-		n.numCalls = 0
-	}
-
-	return n.lastBool, n.lastErr
+	return n.nextResult, n.nextErr
 }
 
 func (n *multiScanNode) Value() core.Doc {
@@ -373,7 +391,10 @@ func (n *multiScanNode) Kind() string {
 }
 
 func (n *multiScanNode) Close() error {
-	return n.scanNode.Close()
+	n.countAndCall(&n.closeCount, func() {
+		n.closeErr = n.scanNode.Close()
+	})
+	return n.closeErr
 }
 
 func (n *multiScanNode) DocumentMap() *core.DocumentMapping {

--- a/internal/planner/scan.go
+++ b/internal/planner/scan.go
@@ -322,6 +322,10 @@ func (p *Planner) Scan(
 // If we have two readers on our multiScanNode, then
 // we call Next() on the underlying scanNode only
 // once every 2 Next() calls on the multiScan
+//
+// NOTE: calling Init() on multiScanNode is subject to counting as well and as such
+// doesn't not provide idempotency guarantees. Counting is purely for performance
+// reasons and removing it should be safe.
 type multiScanNode struct {
 	scanNode   *scanNode
 	numReaders int
@@ -334,6 +338,10 @@ type multiScanNode struct {
 	err        error
 }
 
+// Init initializes the multiScanNode.
+// NOTE: this function is subject to counting based on the number of readers and as such
+// doesn't not provide idempotency guarantees. Counting is purely for performance
+// reasons and removing it should be safe.
 func (n *multiScanNode) Init() error {
 	n.countAndCall(&n.initCount, func() error {
 		return n.scanNode.Init()

--- a/internal/planner/scan.go
+++ b/internal/planner/scan.go
@@ -331,29 +331,34 @@ type multiScanNode struct {
 	closeCount int
 
 	nextResult bool
-	nextErr    error
-	initErr    error
-	startErr   error
-	closeErr   error
+	err        error
 }
 
 func (n *multiScanNode) Init() error {
-	n.countAndCall(&n.initCount, func() {
-		n.initErr = n.scanNode.Init()
+	n.countAndCall(&n.initCount, func() error {
+		return n.scanNode.Init()
 	})
-	return n.initErr
+	return n.err
 }
 
 func (n *multiScanNode) Start() error {
-	n.countAndCall(&n.startCount, func() {
-		n.startErr = n.scanNode.Start()
+	n.countAndCall(&n.startCount, func() error {
+		return n.scanNode.Start()
 	})
-	return n.startErr
+	return n.err
 }
 
-func (n *multiScanNode) countAndCall(count *int, f func()) {
+// countAndCall keeps track of number of requests to call a given function by checking a
+// function's count.
+// The function is only called when the count is 0.
+// If the count is equal to the number of readers, the count is reset.
+// If the function returns an error, the error is stored in the multiScanNode.
+func (n *multiScanNode) countAndCall(count *int, f func() error) {
 	if *count == 0 {
-		f()
+		err := f()
+		if err != nil {
+			n.err = err
+		}
 	}
 	*count++
 
@@ -367,11 +372,12 @@ func (n *multiScanNode) countAndCall(count *int, f func()) {
 // Next only calls Next() on the underlying
 // scanNode every numReaders.
 func (n *multiScanNode) Next() (bool, error) {
-	n.countAndCall(&n.nextCount, func() {
-		n.nextResult, n.nextErr = n.scanNode.Next()
+	n.countAndCall(&n.nextCount, func() (err error) {
+		n.nextResult, err = n.scanNode.Next()
+		return
 	})
 
-	return n.nextResult, n.nextErr
+	return n.nextResult, n.err
 }
 
 func (n *multiScanNode) Value() core.Doc {
@@ -391,10 +397,10 @@ func (n *multiScanNode) Kind() string {
 }
 
 func (n *multiScanNode) Close() error {
-	n.countAndCall(&n.closeCount, func() {
-		n.closeErr = n.scanNode.Close()
+	n.countAndCall(&n.closeCount, func() error {
+		return n.scanNode.Close()
 	})
-	return n.closeErr
+	return n.err
 }
 
 func (n *multiScanNode) DocumentMap() *core.DocumentMapping {

--- a/tests/integration/index/query_with_composite_inxed_on_relation_test.go
+++ b/tests/integration/index/query_with_composite_inxed_on_relation_test.go
@@ -19,7 +19,7 @@ import (
 // This test was added during https://github.com/sourcenetwork/defradb/issues/2924
 // The issue was that [multiScanNode] that keeps track of how many calls to [Next()] should
 // be made, would call [Init()] and [Start()] every time without tracking, which would result
-// child nodes being initialized and started multiple times, which in turn created
+// in child nodes being initialized and started multiple times, which in turn created
 // index iterators without closing them.
 func TestQueryWithCompositeIndexOnManyToOne_WithMultipleIndexedChildNodes_ShouldSucceed(t *testing.T) {
 	test := testUtils.TestCase{

--- a/tests/integration/index/query_with_composite_inxed_on_relation_test.go
+++ b/tests/integration/index/query_with_composite_inxed_on_relation_test.go
@@ -1,0 +1,116 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package index
+
+import (
+	"testing"
+
+	testUtils "github.com/sourcenetwork/defradb/tests/integration"
+)
+
+// This test was added during https://github.com/sourcenetwork/defradb/issues/2924
+// The issue was that [multiScanNode] that keeps track of how many calls to [Next()] should
+// be made, would call [Init()] and [Start()] every time without tracking, which would result
+// child nodes being initialized and started multiple times, which in turn created
+// index iterators without closing them.
+func TestQueryWithCompositeIndexOnManyToOne_WithMultipleIndexedChildNodes_ShouldSucceed(t *testing.T) {
+	test := testUtils.TestCase{
+		Actions: []any{
+			testUtils.SchemaUpdate{
+				Schema: `
+					type User {
+						name: String 
+						devices: [Device]
+					}
+
+					type Device @index(fields: ["owner_id", "manufacturer_id"]) {
+						model: String
+						owner: User 
+						manufacturer: Manufacturer 
+					}
+
+					type Manufacturer {
+						name: String
+						devices: [Device]
+					}
+				`,
+			},
+			testUtils.CreateDoc{
+				DocMap: map[string]any{
+					"name": "John",
+				},
+			},
+			testUtils.CreateDoc{
+				CollectionID: 2,
+				DocMap: map[string]any{
+					"name": "Apple",
+				},
+			},
+			testUtils.CreateDoc{
+				CollectionID: 2,
+				DocMap: map[string]any{
+					"name": "Sony",
+				},
+			},
+			testUtils.CreateDoc{
+				CollectionID: 1,
+				DocMap: map[string]any{
+					"model":        "MacBook Pro",
+					"owner":        testUtils.NewDocIndex(0, 0),
+					"manufacturer": testUtils.NewDocIndex(2, 0),
+				},
+			},
+			testUtils.CreateDoc{
+				CollectionID: 1,
+				DocMap: map[string]any{
+					"model":        "PlayStation 5",
+					"owner":        testUtils.NewDocIndex(0, 0),
+					"manufacturer": testUtils.NewDocIndex(2, 1),
+				},
+			},
+			testUtils.Request{
+				Request: `query {
+					User {
+						devices {
+							model
+							owner {
+								name
+							}
+							manufacturer {
+								name
+							}
+						}
+					}
+				}`,
+				Results: map[string]any{
+					"User": []map[string]any{
+						{
+							"devices": []map[string]any{
+								{
+									"model":        "MacBook Pro",
+									"owner":        map[string]any{"name": "John"},
+									"manufacturer": map[string]any{"name": "Apple"},
+								},
+								{
+									"model":        "PlayStation 5",
+									"owner":        map[string]any{"name": "John"},
+									"manufacturer": map[string]any{"name": "Sony"},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	testUtils.ExecuteTestCase(t, test)
+}

